### ISSUE TITLE
Failed to build OpenSSL on NDK

### ISSUE THREAD

**Issue by olliwang** [ruslo/hunter#1535]
Hunter failed to build OpenSSL on NDK (r18).

      [hunter ** FATAL ERROR **] Build step failed (dir: /Users/usrn/.hunter/_Base/b732407/4e02b95/527d8b0/Build/OpenSSL
      [hunter ** FATAL ERROR **] [Directory:/Users/usrn/.hunter/_Base/Download/Hunter/0.23.11/b732407/Unpacked/cmake/projects/OpenSSL]
      ------------------------------ WIKI -------------------------------
          https://github.com/ruslo/hunter/wiki/error.external.build.failed
      -------------------------------------------------------------------
      CMake Error at /Users/usrn/.hunter/_Base/Download/Hunter/0.23.11/b732407/Unpacked/cmake/modules/hunter_wiki.cmake:12 (message):
      Call Stack (most recent call first):
        /Users/usrn/.hunter/_Base/Download/Hunter/0.23.11/b732407/Unpacked/cmake/modules/hunter_fatal_error.cmake:20 (hunter_wiki)

Then I tried to run `cmake --build Build/` and got something like this:

        LD_LIBRARY_PATH=.: /Users/usrn/Library/Android/sdk/ndk-bundle/toolchains/llvm/prebuilt/darwin-x86_64/bin/clang -DDSO_DLFCN -DHAVE_DLFCN_H -DNDEBUG -DOPENSSL_THREADS -DOPENSSL_NO_DYNAMIC_ENGINE -DOPENSSL_PIC -DOPENSSLDIR="/Users/usrn/.hunter/_Base/b732407/4e02b95/527d8b0/Build/OpenSSL/Install/ssl" -DENGINESDIR="/Users/usrn/.hunter/_Base/b732407/4e02b95/527d8b0/Build/OpenSSL/Install/lib/engines-1.1" -Wall -O3 -pthread -isystem /Users/usrn/Library/Android/sdk/ndk-bundle/sysroot/usr/include/aarch64-linux-android -g -DANDROID -ffunction-sections -funwind-tables -fstack-protector-strong -no-canonical-prefixes -Wa,--noexecstack -Wformat -Werror=format-security --sysroot=ccc -isystem /Users/usrn/Library/Android/sdk/ndk-bundle/sysroot/usr/include -isystem /Users/usrn/Library/Android/sdk/ndk-bundle/sysroot/usr/include/aarch64-linux-android --target=aarch64-none-linux-android24 --gcc-toolchain=/Users/usrn/Library/Android/sdk/ndk-bundle/toolchains/aarch64-linux-android-4.9/prebuilt/darwin-x86_64 -fPIC -o apps/openssl apps/app_rand.o apps/apps.o apps/asn1pars.o apps/ca.o apps/ciphers.o apps/cms.o apps/crl.o apps/crl2p7.o apps/dgst.o apps/dhparam.o apps/dsa.o apps/dsaparam.o apps/ec.o apps/ecparam.o apps/enc.o apps/engine.o apps/errstr.o apps/gendsa.o apps/genpkey.o apps/genrsa.o apps/nseq.o apps/ocsp.o apps/openssl.o apps/opt.o apps/passwd.o apps/pkcs12.o apps/pkcs7.o apps/pkcs8.o apps/pkey.o apps/pkeyparam.o apps/pkeyutl.o apps/prime.o apps/rand.o apps/rehash.o apps/req.o apps/rsa.o apps/rsautl.o apps/s_cb.o apps/s_client.o apps/s_server.o apps/s_socket.o apps/s_time.o apps/sess_id.o apps/smime.o apps/speed.o apps/spkac.o apps/srp.o apps/ts.o apps/verify.o apps/version.o apps/x509.o -L. -lssl -L. -lcrypto -ldl -pthread -latomic --sysroot=/Users/usrn/Library/Android/sdk/ndk-bundle/sysroot
        clang: warning: argument unused during compilation: '-Wa,--noexecstack' [-Wunused-command-line-argument]
        /Users/usrn/Library/Android/sdk/ndk-bundle/toolchains/aarch64-linux-android-4.9/prebuilt/darwin-x86_64/lib/gcc/aarch64-linux-android/4.9.x/../../../../aarch64-linux-android/bin/ld: cannot find crtbegin_dynamic.o: No such file or directory
        /Users/usrn/Library/Android/sdk/ndk-bundle/toolchains/aarch64-linux-android-4.9/prebuilt/darwin-x86_64/lib/gcc/aarch64-linux-android/4.9.x/../../../../aarch64-linux-android/bin/ld: cannot find -ldl
        /Users/usrn/Library/Android/sdk/ndk-bundle/toolchains/aarch64-linux-android-4.9/prebuilt/darwin-x86_64/lib/gcc/aarch64-linux-android/4.9.x/../../../../aarch64-linux-android/bin/ld: cannot find -ldl
        /Users/usrn/Library/Android/sdk/ndk-bundle/toolchains/aarch64-linux-android-4.9/prebuilt/darwin-x86_64/lib/gcc/aarch64-linux-android/4.9.x/../../../../aarch64-linux-android/bin/ld: cannot find -lc
        /Users/usrn/Library/Android/sdk/ndk-bundle/toolchains/aarch64-linux-android-4.9/prebuilt/darwin-x86_64/lib/gcc/aarch64-linux-android/4.9.x/../../../../aarch64-linux-android/bin/ld: cannot find -ldl
        /Users/usrn/Library/Android/sdk/ndk-bundle/toolchains/aarch64-linux-android-4.9/prebuilt/darwin-x86_64/lib/gcc/aarch64-linux-android/4.9.x/../../../../aarch64-linux-android/bin/ld: cannot find crtend_android.o: No such file or directory
        clang: error: linker command failed with exit code 1 (use -v to see invocation)

According the experience of fixing `libxml` on NDK (https://github.com/hunter-packages/libxml2/issues/2). I'm sure the `--sysroot` is wrong again.

Then I tried to print `CMAKE_SYSROOT` and `CMAKE_SYSROOT_COMPILE`. Both points to the same `"/Users/usrn/Library/Android/sdk/ndk-bundle/sysroot"`, and that's wrong. It should be something like `"/Users/usrn/Library/Android/sdk/ndk-bundle/platforms/android-24/arch-arm64"`.

The last `--sysroot` in the last executed command comes from the `EX_LIBS` variable defined in `Makefile`.

        ....
        PLIB_LDFLAGS= 
        EX_LIBS= -ldl -pthread -latomic --sysroot=/Users/usrn/Library/Android/sdk/ndk-bundle/sysroot 
        LIB_CFLAGS=-fPIC -DOPENSSL_USE_NODELETE
        ....

However, it seems the wrong `--sysroot` doesn't come from `hunter`. Because I tried set all `--sysroot` to something else in `hunter_dump_cmake_flags.cmake`. And the `EX_LIBS` never changes. 

I also tried to change the `--sysroot` value to `"/Users/usrn/Library/Android/sdk/ndk-bundle/platforms/android-24/arch-arm64"` (in EX_LIBS) and executed `make` manually, and it passed the building process.

So the question is, how can I fix `hunter` to set the `--sysroot` value correctly? Thanks.

**Comment by ruslo** [comment]
> Both points to the same "/Users/usrn/Library/Android/sdk/ndk-bundle/sysroot", and that's wrong. It should be something like "/Users/usrn/Library/Android/sdk/ndk-bundle/platforms/android-24/arch-arm64"

Are you using CMake-based NDK toolchain or Android Studio toolchain? According to this comment the sysroot value you're looking for is in `CMAKE_EXE_LINKER_FLAGS` variable:
* https://github.com/hunter-packages/libxml2/issues/2#issuecomment-410318743

> So the question is, how can I fix hunter to set the --sysroot value correctly? Thanks

OpenSSL use custom build scheme, so the first question is how OpenSSL should be build with NDK r18.

**Comment by ruslo** [comment]
> Not sure where to pass the --sysroot parameter to configure in hunter

OpenSSL build schemes:
* https://github.com/ruslo/hunter/blob/305e1831180a1f80128c5f0e9a0e10b8b928342b/cmake/projects/OpenSSL/hunter.cmake#L387-L403

`CMAKE_EXE_LINKER_FLAGS` will be available after this command:
* https://github.com/ruslo/hunter/blob/305e1831180a1f80128c5f0e9a0e10b8b928342b/cmake/projects/OpenSSL/schemes/url_sha1_openssl.cmake.in#L58-L60

See:
* https://github.com/ruslo/hunter/blob/305e1831180a1f80128c5f0e9a0e10b8b928342b/cmake/modules/hunter_dump_cmake_flags.cmake#L62-L75

**Comment by olliwang** [comment]
@ruslo  Thanks for the tips. I can make it work now by replace 

https://github.com/ruslo/hunter/blob/305e1831180a1f80128c5f0e9a0e10b8b928342b/cmake/projects/OpenSSL/schemes/url_sha1_openssl.cmake.in#L104

to 

    STRING(REGEX REPLACE ".*[ ]--sysroot[ ]([A-Za-z0-9/-]+)[ ]?.*" "\\1" 
           SYSROOT ${CMAKE_EXE_LINKER_FLAGS})
    if (${SYSROOT} STREQUAL ${CMAKE_EXE_LINKER_FLAGS})
      set(SYSROOT ${CMAKE_SYSROOT})
    endif()
    list(APPEND configure_opts "-latomic --sysroot=${SYSROOT}")

Basically I tried to extract the `sysroot` from  `CMAKE_EXE_LINKER_FLAGS` and fallback to `CMAKE_SYSROOT` if failed. Is that ok to submit a PR?

**Comment by olliwang** [comment]
@ruslo Just tried. It's working, too. 

    list(APPEND configure_opts "-latomic ${CMAKE_EXE_LINKER_FLAGS}")

Since the fix is quite simple, I think it's faster for you to fix it instead of a PR. What do you think?

**Comment by ruslo** [comment]
> Since the fix is quite simple, I think it's faster for you to fix it instead of a PR. What do you think?

Don't see the difference, I will run all tests anyway.

**Comment by olliwang** [comment]
BTW, I tested the modified code by running

    cmake --build app/.externalNativeBuild/cmake/debug/arm64-v8a

So Android Studio won't overwrite the scheme file.

**Comment by olliwang** [comment]
This is how I added OpenSSL

    hunter_add_package(OpenSSL)
    find_package(OpenSSL REQUIRED)
    target_link_libraries(moui-app PRIVATE OpenSSL::SSL OpenSSL::Crypto)

**Comment by ruslo** [comment]
Fix available in  https://github.com/ruslo/hunter/releases/tag/v0.23.25

**Comment by olliwang** [comment]
Works perfectly. Thanks.
